### ISSUE TITLE
A new rule proposal to make const variable values immutable

### ISSUE THREAD

**Issue by Nirvik-D** [eslint/eslint#10442]
**Please describe what the rule should do:**
The new rule should handle the immutability of values of const variables to a deeper level. Although, there is an existing library called deep-freeze, which works recursively, the rule should do the job faster than running a third party library.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[ ] Warns about a potential error
[X] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
const obj = Object.freeze({ foo: {} });
obj.bar = 123
// TypeError: Can't add property bar, object is not extensible
obj.foo = {}
// TypeError: Cannot assign to read only property 'foo' of #<Object>

// Object is not immutable
obj.foo.qux = 'abc';
obj.foo.qux // prints abc

// Rule should make values of const variables immutable
obj.foo.qux = 'abc';
obj.foo.qux // TypeError: Cannot assign to read only property 'qux' of #<Object>
```

**Why should this rule be included in ESLint (instead of a plugin)?**
To remove dependencies from a third party library.

**Comment by platinumazure** [comment]
I'm :-1: on including this in core, as it could give users an incorrect understanding of how `const` actually works in JavaScript. (`block-scoped-var` already causes similar confusion about hoisting of `var` variables.)

That said, I think this would make a great idea for a plugin!

**Comment by ljharb** [comment]
`const` means “constant reference”, it does not mean “constant value”, so i heartily agree with https://github.com/eslint/eslint/issues/10442#issuecomment-394497244

**Comment by g-plane** [comment]
You can use [eslint-plugin-immutable](https://github.com/jhusain/eslint-plugin-immutable)
